### PR TITLE
Add an option to include comments in FTA score calculation

### PR DESCRIPTION
**Pull request by bencollins54**
Add an option to include comments in FTA score calculation.
Exclude comments by default.

### CODE REVIEW

File: crates/fta/src/parse/mod.rs [hunk 1]
```diff
@@ -67,14 +68,14 @@ struct CountingComments {
...
     fn add_leading(self: &CountingComments, _pos: BytePos, _comment: Comment) {
         let current_count = self.count.get();
-        println!("Comment: {:?}", _comment.text);
+        debug!("Comment: {:?}", _comment.text);
         self.count
             .set(current_count + 1 + _comment.text.matches('\n').count());
     }
 
     fn add_leading_comments(self: &CountingComments, _pos: BytePos, _comments: Vec<Comment>) {
         _comments.iter().for_each(|comment| {
```
Comment (sgb-io) [inline crates/fta/src/parse/mod.rs:77]: Now that the feature is working with independent tests, IMO, these debug statements should just be removed. There's no value in keeping them there vs re-adding if there is something to diagnose.

File: crates/fta/src/lib.rs [hunk 2]
```diff
@@ -19,7 +19,7 @@ use swc_ecma_parser::error::Error;
 use utils::{check_score_cap_breach, get_assessment, is_valid_file, warn_about_language};
 
 pub fn analyze_file(module: &Module, line_count: usize) -> (usize, HalsteadMetrics, f64) {
-    let cyclo = cyclo::cyclomatic_complexity(module.clone());
+    let cyclo = cyclo::cyclomatic_complexity(module);
```
Comment (sgb-io) [inline crates/fta/src/lib.rs:22]: What's the purpose/impact of this change?

File: crates/fta/src/lib.rs [hunk 3]
```diff
@@ -92,7 +94,11 @@ fn do_analysis(
     source_code: &str,
     use_tsx: bool,
 ) -> Result<FileData, Error> {
-    let (result, line_count) = parse::parse_module(source_code, use_tsx);
+    let (result, line_count) = parse::parse_module(
+        source_code,
+        use_tsx,
+        config.include_comments.unwrap_or(false),
```
Comment (sgb-io) [inline crates/fta/src/lib.rs:100]: The 3rd argument here is pulled inconsistent vs the `use_tsx` option.

It should probably be passed in to be consistent with how `use_tsx` is? Unless there is a counter-argument against such a change

File: crates/fta/src/parse/mod.rs [hunk 4]
```diff
@@ -27,11 +40,109 @@ pub fn parse_module(source: &str, use_tsx: bool) -> (Result<Module, Error>, usiz
...
         EsVersion::Es2020,
         SourceFileInput::from(&*fm),
-        None,
+        Some(&comments),
     );
 
     let mut parser = Parser::new_from(lexer);
     let parsed = parser.parse_module();
 
+    debug!(
```
Comment (sgb-io) [inline crates/fta/src/parse/mod.rs:49]: Debug statement should be removed imo

File: crates/fta/src/parse/mod.rs [hunk 5]
```diff
@@ -27,11 +40,109 @@ pub fn parse_module(source: &str, use_tsx: bool) -> (Result<Module, Error>, usiz
...
+        let current_count = self.count.get();
+        let comment_count: usize = _comments
+            .iter()
+            .map(|comment| comment.text.matches('\n').count())
+            .sum();
+        self.count.set(current_count + 1 + comment_count);
+    }
+
+    fn add_trailing(self: &CountingComments, _pos: BytePos, _comment: Comment) {
+        //         let current_count = self.count.get();
```
Comment (sgb-io) [inline crates/fta/src/parse/mod.rs:89]: Remove commented out lines

File: crates/fta/src/lib.rs (hunk 2 above)
Comment (bencollins54) [reply crates/fta/src/lib.rs:22]: Generate a flame graph and @marcusedwardhaslam noticed this.
I don't have the exact numbers still (I want to say it was 55 vs 60 or around 8%  for one of the large repos) but it was a small but consistent speed increase.

File: crates/fta/src/parse/mod.rs [hunk 6]
```diff
@@ -27,11 +40,109 @@ pub fn parse_module(source: &str, use_tsx: bool) -> (Result<Module, Error>, usiz
...
+    fn take_leading(self: &CountingComments, _pos: BytePos) -> Option<Vec<Comment>> {
+        None
+    }
+
+    fn take_trailing(self: &CountingComments, _pos: BytePos) -> Option<Vec<Comment>> {
+        None
+    }
+
+    fn move_leading(&self, _from: swc_common::BytePos, _to: swc_common::BytePos) {
+        todo!()
```
Comment (sgb-io) [inline crates/fta/src/parse/mod.rs:115]: What is the effect of these `todo!()` statements?
Is it visable to remove these un-implemented methods entirely?

Not a dealbreaker; just seeking tidy-up opportunities

File: crates/fta/src/parse/mod.rs [hunk 7]
```diff
@@ -27,11 +40,109 @@ pub fn parse_module(source: &str, use_tsx: bool) -> (Result<Module, Error>, usiz
...
+    };
+
     (parsed, line_count)
 }
+
+struct CountingComments {
+    count: Cell<usize>,
+}
+
+impl Comments for CountingComments {
```
Comment (sgb-io) [inline crates/fta/src/parse/mod.rs:68]: You asked about this: I actually think this implementation can just stay in this file. It's not overly large and it's all relevant to the `parse` module

File: crates/fta/src/parse/mod.rs (hunk 1 above)
Comment (bencollins54) [reply crates/fta/src/parse/mod.rs:77]: Do we want to remove the other `debug` outputs i.e. `crates/fta/src/halstead/mod.rs`

Comment (sgb-io) [reply crates/fta/src/parse/mod.rs:77]: I think I've answered this in other review comments but I think, basically: remove them all relating to counting comments. This is an internal feature that is now verified via tests.

Debug statements are useful for end users to get additional output if they have a problem, so they should be limited to that scope, imo. 

FYI, the top level `lib.rs` checks for `DEBUG` in the env (might not always need to use the proper Rust flag)

File: crates/fta/src/lib.rs (hunk 2 above)
Comment (sgb-io) [reply crates/fta/src/lib.rs:22]: Cool, assuming this PR is green i.e. nothing is broke and everything looks fast, sounds good to me 👍

File: crates/fta/src/parse/mod.rs (hunk 6 above)
Comment (bencollins54) [reply crates/fta/src/parse/mod.rs:115]: The `todo!` were auto-generated but can probably be removed.
I think the un-implemented methods are required due to the trait but I'll double check.

File: crates/fta/src/lib.rs (hunk 3 above)
Comment (bencollins54) [reply crates/fta/src/lib.rs:100]: `use_tsx` isn't part of config and other values from config are used within this method.

File: crates/fta/src/parse/mod.rs (hunk 7 above)
Comment (bencollins54) [reply crates/fta/src/parse/mod.rs]: Fine :+1:

File: crates/fta/src/config/mod.rs [hunk 8]
```diff
@@ -57,6 +58,9 @@ pub fn read_config(config_path: &str) -> FtaConfig {
             },
             output_limit: provided_config.output_limit.or(default_config.output_limit),
             score_cap: provided_config.score_cap.or(default_config.score_cap),
+            include_comments: provided_config
```
[the remark this answers is not in the document]
Comment (bencollins54) [reply crates/fta/src/config/mod.rs:61]: Added a test for this.

**Comment by sgb-io** [comment]
e2e test case in `test.yml` needs the line count updating

**Comment by sgb-io** [comment]
Fta score in the e2e test also needs to be updated, since the number of lines was corrected

**Pull request merged**